socorro/scripts/fetch_crashids.py

```diff
@@ -29,17 +32,95 @@
+
+    NOTE(willkg): There are multiple infinities and not all infinities are equal, so what we're
+    doing here is wrong, but it's helpful. We can rename it if someone gets really annoyed.
+
+    """
+    def __eq__(self, obj):
+        return isinstance(obj, Infinity)
+
+    def __lt__(self, obj):
+        return False
+
+    def __repr__(self):
+        return 'Infinity'
+
+    def __sub__(self, obj):
+        if isinstance(obj, Infinity):
+            return 0
+        return self
+
+    def __rsub__(self, obj):
+        # We don't need to deal with negative infinities, so let's not
+        raise ValueError('This Infinity does not support right-hand-side')
+
+
+# For our purposes, there is only one infinity
+INFINITY = Infinity()
+
+
+def fetch_crashids(params, num_results):
+    """Generator that returns crash ids
+
+    :arg dict params: dict of super search parameters to base the query on
+    :arg varies num: number of results to get or INFINITY
+
+    :returns: generator of crash ids
+
+    """
     url = HOST + '/api/SuperSearch/'
 
-    resp = requests.get(url, params)
-    if resp.status_code == 200:
+    # Set up retrying on 429s
+    retries = Retry(total=32, backoff_factor=1, status_forcelist=[429])
+    session = requests.Session()
+    session.mount('http://', HTTPAdapter(max_retries=retries))
+    session.mount('https://', HTTPAdapter(max_retries=retries))
+
+    # Set up first page
+    params['_results_offset'] = 0
+    params['_results_number'] = min(MAX_PAGE, num_results)
+
+    # Fetch pages of crash ids until we've gotten as many as we want or there aren't any more to get
+    crashids_count = 0
+    while True:
+        resp = session.get(url, params=params)
+        if resp.status_code != 200:
+            raise Exception('Bad response: %s %s' % (resp.status_code, resp.content))
+
         hits = resp.json()['hits']
-        crashids = [hit['uuid'] for hit in hits]
-        return crashids
 
-    raise Exception('Bad response: %s %s' % (resp.status_code, resp.content))
+        for hit in hits:
+            crashids_count += 1
+            yield hit['uuid']
+
+            # If we've gotten as many crashids as we need, we return
+            if crashids_count >= num_results:
+                return
+
+        # If there are no more crash ids to get, we return
+        total = resp.json()['total']
+        if not hits or crashids_count >= total:
+            return
+
+        # Get the next page, but only as many results as we need
+        params['_results_offset'] += MAX_PAGE
+        params['_results_number'] = min(
+            # MAX_PAGE is the maximum we can request
+            MAX_PAGE,
+
+            # The number of results Super Search can return to us that is hasn't returned so far
+            total - crashids_count,
+
+            # The numver of results we want that we haven't gotten, yet
+            num_results - crashids_count
+        )
```

Comment:
This works in the non-all case because it's just numbers.

This works in the `all` case, because `num_results` is `INFINITY` and we've coded `INFINITY` to return `INFINITY` when in the left-hand-side of a subtraction operation so that number will always be bigger than the other two.